src/components/3d_editor/advanced_geometry/python_transformation/PythonTransformation.tsx

```diff
@@ -0,0 +1,174 @@
+
+interface PythonTransformationProps {
+    materials: any[];
+    show: boolean;
+    onSubmit: (newMaterials: any[]) => void;
+    onHide: () => void;
+}
+
+interface PythonTransformationState {
+    materials: any[];
+    selectedMaterials: any[];
+    isLoading: boolean;
+    isRunning: boolean;
+    pyodide: any;
+    pythonCode: string;
+    pythonOutput: string;
+}
+
+class PythonTransformation extends React.Component<
+    PythonTransformationProps,
+    PythonTransformationState
+> {
+    constructor(props: PythonTransformationProps) {
+        super(props);
+        this.state = {
+            materials: props.materials,
+            selectedMaterials: [props.materials[0]],
+            isLoading: true,
+            isRunning: false,
+            pyodide: null,
+            pythonCode: "",
+            pythonOutput: "",
+        };
+    }
+
+    // eslint-disable-next-line no-unused-vars
+    componentDidUpdate(prevProps: PythonTransformationProps, prevState: any) {
+        const { materials } = this.props;
+        if (prevProps.materials !== materials) {
+            // eslint-disable-next-line react/no-did-update-set-state
+            this.setState({ materials });
+        }
+    }
+
+    onPyodideLoad = (pyodideInstance: any) => {
+        this.setState({ pyodide: pyodideInstance, isLoading: false }, async () => {
+            // redirecting stdout for print and errors per https://pyodide.org/en/stable/usage/streams.html
+            pyodideInstance.setStdout({
+                batched: (text: string) => this.redirectPyodideStdout(text),
+            });
+        });
+    };
+
+    redirectPyodideStdout = (text: string) => {
+        this.setState((prevState) => ({
+            pythonOutput: prevState.pythonOutput + text + "\n",
+        }));
+    };
+
+    runPythonCode = async () => {
+        const { pyodide, pythonCode } = this.state;
+        this.setState({ pythonOutput: "" });
+
+        try {
+            await pyodide.runPythonAsync(pythonCode);
+        } catch (error: any) {
+            this.setState({ pythonOutput: error.message });
+        }
+    };
+
+    handleRun = async () => {
+        this.setState({ isRunning: true });
+        try {
+            await this.runPythonCode();
+        } catch (error: any) {
+            NPMsAlert.error(error.message);
+        } finally {
+            this.setState({ isRunning: false });
+        }
+    };
+
+    handleSubmit = () => {
+        // TODO: return newly created materials in the next task
+        // eslint-disable-next-line react/destructuring-assignment
+        this.props.onSubmit(this.state.selectedMaterials);
+    };
+
+    render() {
+        const { isLoading, isRunning, pythonCode, pythonOutput, materials, selectedMaterials } =
+            this.state;
+        const { show, onHide } = this.props;
+
+        return (
+            // TODO: fix DarkMaterialUITheme in cove.js and remove ThemeProvider
+            <ThemeProvider theme={theme}>
```

Comment:
Why do we need theme provider?

Reply:
I've explained before -- workaround because Dark theme that the rest of MD has is not eqiuvalent to Light theme -- it breaks the app.